### PR TITLE
refactoring: replace a byte.Buffer with a byte slice

### PR DESCRIPTION
**Pull request by hajimehoshi**
This change also removes `length` member with the len of the slice.

This slightly make the audio decoding performance worse, but I am not sure why.

```
goos: darwin
goarch: arm64
pkg: github.com/gen2brain/mpeg
               │   old.txt   │              new.txt               │
               │   sec/op    │   sec/op     vs base               │
DecodeVideo-12   76.90µ ± 1%   76.34µ ± 1%  -0.73% (p=0.024 n=10)
DecodeAudio-12   47.64µ ± 1%   49.36µ ± 2%  +3.62% (p=0.002 n=10)
RGBA-12          41.29µ ± 0%   40.53µ ± 0%  -1.83% (p=0.001 n=10)
geomean          53.28µ        53.45µ       +0.32%

               │  old.txt   │              new.txt              │
               │    B/op    │    B/op     vs base               │
DecodeVideo-12   76.00 ± 1%   75.00 ± 0%  -1.32% (p=0.008 n=10)
DecodeAudio-12   46.50 ± 1%   48.00 ± 2%  +3.23% (p=0.000 n=10)
RGBA-12          40.00 ± 2%   40.00 ± 0%       ~ (p=0.582 n=10)
geomean          52.09        52.41       +0.62%

               │   old.txt    │               new.txt               │
               │  allocs/op   │ allocs/op   vs base                 │
DecodeVideo-12   0.000 ± 0%     0.000 ± 0%       ~ (p=1.000 n=10) ¹
DecodeAudio-12   0.000 ± 0%     0.000 ± 0%       ~ (p=1.000 n=10) ¹
RGBA-12          0.000 ± 0%     0.000 ± 0%       ~ (p=1.000 n=10) ¹
geomean                     ²               +0.00%                ²
¹ all samples are equal
² summaries must be >0 to compute geomean
```

### CODE REVIEW

File: buffer.go [hunk 1]
```diff
@@ -189,41 +179,40 @@ func (b *Buffer) tell() int {
...
 	}
 
 	return b.bitIndex >> 3
 }
 
 func (b *Buffer) discardReadBytes() {
 	bytePos := b.bitIndex >> 3
-	if bytePos == b.length {
-		b.bytes.Truncate(0)
+	if bytePos == len(b.bytes) {
```
Comment (hajimehoshi) [inline buffer.go:190]: Is this logic correct by the way? What if bitIndex is 9 and length (len(b.bytes)) is 1? Isn't 1bit discarded unexpectedly?

Comment (gen2brain) [reply buffer.go:190]: The logic is rewritten from https://github.com/phoboslab/pl_mpeg/blob/master/pl_mpeg.h#L1553.

Comment (hajimehoshi) [reply buffer.go:190]: So, is the original logic correct...? This seems a little suspicious

Comment (hajimehoshi) [reply buffer.go:190]: Aha, now I think this is correct: when `bytePos == len(b.bytes)`, the lower 3bits of `bitIndex` must be 0, or `bitIndex` would exceed the buffer length.

**Comment by gen2brain** [comment]
I remember I switched back and forth between byte slice and bytes.Buffer, I can't remember what was the difference and why bytes.Buffer is finally used.

**Comment by hajimehoshi** [comment]
From my intuition, the current byte.Buffer just adds overheads, but I might be wrong.

**Comment by gen2brain** [comment]
Merged, thanks.

**Pull request merged**